cpp/src/io/parquet/decode_fixed.cu

```diff
@@ -894,32 +894,27 @@ __device__ inline bool maybe_has_nulls(page_state_s* s)
   return run_val != s->col.max_level[lvl];
 }
 
-template <int decode_block_size_t, typename state_buf>
-inline __device__ void bool_plain_decode(page_state_s* s, state_buf* sb, int t, int to_decode)
+template <typename state_buf, typename thread_group>
+inline __device__ void bool_plain_decode(page_state_s* s,
+                                         state_buf* sb,
+                                         int target_pos,
+                                         thread_group const& group)
 {
-  int pos              = s->dict_pos;
-  int const target_pos = pos + to_decode;
-  __syncthreads();  // Make sure all threads have read dict_pos before it changes at the end.
+  int const pos = s->dict_pos;
+  int const t   = group.thread_rank();
+  // Ensure all threads have the dict_pos
+  group.sync();
 
-  while (pos < target_pos) {
-    int const batch_len = min(target_pos - pos, decode_block_size_t);
+  for (auto bit_pos = pos + t; bit_pos < target_pos; bit_pos += group.size()) {
+    int const byte_offset       = bit_pos >> 3;
+    int const bit_in_byte_index = bit_pos & 7;
 
-    if (t < batch_len) {
-      int const bit_pos           = pos + t;
-      int const byte_offset       = bit_pos >> 3;
-      int const bit_in_byte_index = bit_pos & 7;
+    uint8_t const* const read_from = s->data_start + byte_offset;
+    bool const read_bit            = (*read_from) & (1 << bit_in_byte_index);
 
-      uint8_t const* const read_from = s->data_start + byte_offset;
-      bool const read_bit            = (*read_from) & (1 << bit_in_byte_index);
-
-      int const write_to_index     = rolling_index<state_buf::dict_buf_size>(bit_pos);
-      sb->dict_idx[write_to_index] = read_bit;
-    }
-
-    pos += batch_len;
+    int const write_to_index     = rolling_index<state_buf::dict_buf_size>(bit_pos);
+    sb->dict_idx[write_to_index] = read_bit;
   }
-
-  if (t == 0) { s->dict_pos = pos; }
```

Comment:
Simply moving this one outside the function at L1235 (and keeping all original changes) also just fixes the said problem.